### PR TITLE
Reactive forms

### PR DESCRIPTION
**Pull request by N1sMyName**

### CODE REVIEW

File: src/app/static-nav-top/search-bar/search-bar.component.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    if (this.searchQuery.errors) {
+      this.errorStyles['isError'] = true;
+      this.errorMsg = 'Required min-length 3';
+      this.dropDownResult.hidden = true;
+      return;
+    }
+    // make copy of data
+    this.list = [...JSON.parse(JSON.stringify(this.data))];
+
+    if (this.list) {
```
Comment (serhii-lisniakov) [inline src/app/static-nav-top/search-bar/search-bar.component.ts:88]: this is useless if!

File: src/app/static-nav-top/search-bar/search-bar.component.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  closeResultWindow() {
+    this.dropDownResult.hidden = true;
+  }
+  moveToProduct(id: number) {
+    const link = `/products/${id}`;
+    this.router.navigateByUrl(link);
+    this.closeResultWindow();
+  }
+}
+[];
```
Comment (serhii-lisniakov) [inline src/app/static-nav-top/search-bar/search-bar.component.ts:112]: what is this?)

File: src/app/static-nav-top/search-bar/search-bar.component.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    'error-container': true,
+    isError: false,
+  };
+
+  ngOnInit(): void {
+    this.api.getProducts().subscribe((res) => {
+      this.data = res;
+    });
+
+    document.addEventListener('keydown', this.handleCloseByEsc.bind(this));
```
Comment (serhii-lisniakov) [inline src/app/static-nav-top/search-bar/search-bar.component.ts:43]: if you still decided to use native event listener please, use ElementRef injection in order not to shake document itself

File: src/app/static-nav-top/search-bar/search-bar.component.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  ngOnDestroy() {
+    document.removeEventListener('keydown', this.handleCloseByEsc);
+  }
+  handleCloseByEsc(event: KeyboardEvent) {
+    if (event.key === 'Escape') {
+      this.dropDownResult.hidden = true;
+      this.errorStyles.isError = false;
+    }
+  }
+  handleSearchByEnter(event: KeyboardEvent) {
```
Comment (serhii-lisniakov) [inline src/app/static-nav-top/search-bar/search-bar.component.ts:55]: where do you use this?

File: src/app/static-nav-top/search-bar/search-bar.component.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    hidden: true,
+    'search-result': true,
+  };
+  public errorStyles = {
+    'error-container': true,
+    isError: false,
+  };
+
+  ngOnInit(): void {
+    this.api.getProducts().subscribe((res) => {
```
Comment (serhii-lisniakov) [inline src/app/static-nav-top/search-bar/search-bar.component.ts:39]: spread it to separate method, try to keep in Oninit body only method calls

File: src/app/static-nav-top/search-bar/search-bar.component.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+      this.data = res;
+    });
```
Comment (serhii-lisniakov) [inline src/app/static-nav-top/search-bar/search-bar.component.ts:41]: spread it to separate method, try to keep in Oninit body only method calls

File: src/app/static-nav-top/search-bar/search-bar.component.ts (hunk 3 above)
Comment (serhii-lisniakov) [reply src/app/static-nav-top/search-bar/search-bar.component.ts:43]: and spread it as well

File: src/app/static-nav-top/search-bar/search-bar.component.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+  selector: 'app-search-bar',
+  templateUrl: './search-bar.component.html',
+  styleUrls: ['./search-bar.component.sass'],
+})
+export class SearchBarComponent {
+  constructor(
+    private api: MimicrestService,
+    private router: Router,
+    private fb: FormBuilder
+  ) {}
```
Comment (serhii-lisniakov) [inline src/app/static-nav-top/search-bar/search-bar.component.ts:17]: keep constructor under variables declarations

File: src/app/Services/mimicrest.service.ts [hunk 8]
```diff
@@ -1,16 +1,17 @@
 import { HttpClient, HttpErrorResponse } from '@angular/common/http';
 import { Injectable } from '@angular/core';
-import { Observable, throwError } from 'rxjs';
+import { Observable, throwError, throwIfEmpty } from 'rxjs';
```
Comment (serhii-lisniakov) [inline src/app/Services/mimicrest.service.ts:3]: remove unused imports

File: src/app/static-nav-top/search-bar/search-bar.component.ts (hunk 2 above)
Comment (N1sMyName) [reply src/app/static-nav-top/search-bar/search-bar.component.ts:112]: function hide bar output search when redirect, and attached to event in the end of same bar as alternative for closing.
image

**Pull request merged**